test/indexed_dataset_test.py

```diff
@@ -0,0 +1,631 @@
+        ds = ds.cache()
+        n = tf.compat.v1.data.make_one_shot_iterator(ds).get_next()
+    res = run_steps(g, n, 5)
+    assert res == [0,1,2,3,4]
+    res = run_steps(g, n, 4)
+    assert res == [5,6,7,8]
+    res = run_steps(g, n, 1)
+    assert res == [9]
+    with pytest.raises(tf.errors.OutOfRangeError):
+        res = run_steps(g, n, 1)
+
+@initialize_ckpt
+def test_tf_record():
+    pass
+
+@initialize_ckpt
+def test_take():
+    g = tf.Graph()
+    with g.as_default():
+        ds = tf.data.Dataset.from_tensor_slices(list(range(10)))
+        ds = ds.take(5)
+        n = tf.compat.v1.data.make_one_shot_iterator(ds).get_next()
+    res = run_steps(g, n, 1)
+    assert res == [0]
+    res = run_steps(g, n, 3)
+    assert res == [1,2,3]
+    res = run_steps(g, n, 1)
+    assert res == [4]
+    with pytest.raises(tf.errors.OutOfRangeError):
+        res = run_steps(g, n, 1)
+
+@initialize_ckpt
+def test_tf_record_iterleave():
+    g = tf.Graph()
+    with g.as_default():
+        file_names = [
+            "/cmsdata/ssd1/cmslab/imagenet-data/aws/train-00{}-of-01024"
+            .format(i) for i in range(256, 256+128)
+        ]
+        ds = tf.data.TFRecordDataset.list_files(file_names)
+        ds = ds.interleave(tf.data.TFRecordDataset)
+        iterator = tf.compat.v1.data.make_one_shot_iterator(ds)
+        n = iterator.get_next()
+
+    run_steps(g, n, 1)
+
+@initialize_ckpt
+def test_tf_record_experimental_iterleave():
+    g = tf.Graph()
+    with g.as_default():
+        file_names = [
+            "/cmsdata/ssd1/cmslab/imagenet-data/aws/train-00{}-of-01024".format(i) for i in range(256, 256+128)
+        ]
+        ds = tf.data.TFRecordDataset.list_files(file_names, shuffle=False)
+        ds = ds.apply(
+            tf.data.experimental.parallel_interleave(
+                tf.data.TFRecordDataset, cycle_length=1))
+        iterator = tf.compat.v1.data.make_one_shot_iterator(ds)
+        n = iterator.get_next()
+
+    run_steps(g, n, 10)
+
+@initialize_ckpt
+def test_interleave_and_prefetch():
+    g = tf.Graph()
+    with g.as_default():
+        ds = tf.data.Dataset.from_tensor_slices([[0,1,2],[3,4,5],[6,7,8],
+                                                 [9,10,11],[12,13,14],[15,16,17],
+                                                 [18,19,20],[21,22,23],[24,25,26]])
+        ds = ds.interleave(tf.data.Dataset.from_tensor_slices)
+        ds = ds.prefetch(10)
+        n = tf.compat.v1.data.make_one_shot_iterator(ds).get_next()
+    res = run_steps(g, n, 5)
+    assert res == [0,3,6,9,12]
+    res = run_steps(g, n, 4)
+    assert res == [15,18,21,24]
+    res = run_steps(g, n, 6)
+    assert res == [1,4,7,10,13,16]
+    res = run_steps(g, n, 9)
+    assert res == [19,22,25,2,5,8,11,14,17]
+    res = run_steps(g, n, 3)
+    assert res == [20,23,26]
+    with pytest.raises(tf.errors.OutOfRangeError):
+        res = run_steps(g, n, 1)
+
+@initialize_ckpt
+def test_imagenet():
```

Comment:
Correctness tests when the number of workers and the batch size are changed are necessary.

Reply:
Added more tests including scale in/out situations.